src/extract.c

```diff
@@ -5,48 +5,124 @@
 #include "backports.h"
 #include <string.h>
 
-int find_offset(SEXP x, SEXP index, int i) {
+int find_offset(SEXP x, SEXP index, int i, int strict) {
   if (Rf_length(index) > 1) {
-    Rf_errorcall(R_NilValue, "Index %i must have length 1", i + 1);
+    Rf_errorcall(R_NilValue,
+      "Index %i must have length 1, not %i.",
+      i + 1,
+      Rf_length(index)
+    );
+  } else if (strict && Rf_length(index) == 0) {
+    Rf_errorcall(R_NilValue, "Index %i must have length 1, not 0", i + 1);
   }
 
   int n = Rf_length(x);
-  if (n == 0)
-    return -1;
+  if (n == 0) {
+    if (strict)
+      Rf_errorcall(R_NilValue, "Object being plucked must have at least one element.");
+    else
+      return -1;
+  }
 
   if (TYPEOF(index) == INTSXP) {
     int val = INTEGER(index)[0];
 
-    if (val == NA_INTEGER)
-      return -1;
+    if (val == NA_INTEGER) {
+      if (strict)
+        Rf_errorcall(R_NilValue,
+          "Index %i must be finite, not NA.",
+          i + 1
+        );
+      else
+        return -1;
+    }
 
     val--;
-    if (val < 0 || val >= n)
-      return -1;
+    if (val < 0) {
+      if (strict)
+        Rf_errorcall(R_NilValue,
+          "Index %i must be greater than 0, not %i.",
+          i + 1,
+          val + 1
+        );
+      else
+        return -1;
+    } else if (val >= n) {
+      if (strict)
+        Rf_errorcall(R_NilValue,
+          "Index %i exceeds the length of object being plucked (%i > %i).",
+          i + 1,
+          val + 1,
+          n
+        );
+      else
+        return -1;
+    }
 
     return val;
   } if (TYPEOF(index) == REALSXP) {
     double val = REAL(index)[0];
 
-    if (!R_finite(val))
-      return -1;
+    if (!R_finite(val)) {
+      if (strict)
+        Rf_errorcall(R_NilValue,
+          "Index %i must be finite, not %s.",
+          i + 1,
+          Rf_translateCharUTF8(Rf_asChar(index))
+        );
+      else
+        return -1;
+    }
 
     val--;
-    if (val < 0 || val >= n)
-      return -1;
+    if (val < 0) {
+      if (strict)
+        Rf_errorcall(R_NilValue,
+          "Index %i must be greater than 0, not %.0f.",
+          i + 1,
+          val + 1
+        );
+      else
+        return -1;
+    } else if (val >= n) {
+      if (strict)
+        Rf_errorcall(R_NilValue,
+          "Index %i exceeds the length of object being plucked (%.0f > %i).",
```

Comment:
Why `%.0f` and not `%i`?

Reply:
Because of `double val = REAL(index)[0];`, using `%i` gives `nvalid index (1): index must be greater than 0, not 536870913.`.

Reply:
Ah got it.